### PR TITLE
Add public examples to all functions in std.math

### PR DESCRIPTION
**Pull request by JackStouffer**
Finishes everything except `rndtonl` which is blocked by issue 18693

### CODE REVIEW

File: std/math.d [hunk 1]
```diff
@@ -6220,13 +6309,25 @@ if (isFloatingPoint!(R) && isFloatingPoint!(X))
...
 {
     return copysign(cast(R) to, from);
 }
 
+///
+@safe pure nothrow @nogc unittest
+{
+    assert(copysign(1.0, 1.0).feqrel(1.0) > 16);
+    assert(copysign(1.0, -0.0).feqrel(-1.0) > 16);
+    assert(copysign(1UL, -1.0).feqrel(-1.0) > 16);
```
Comment (wilzbach) [inline std/math.d]: Maybe add `copysign(-1.0, -1.0)` too?

File: std/math.d [hunk 2]
```diff
@@ -6618,6 +6735,12 @@ float nextUp(float x) @trusted pure nothrow @nogc
     return x;
 }
 
+///
+@safe @nogc pure nothrow unittest
+{
+
```
Comment (wilzbach) [inline std/math.d]: ?

File: std/math.d [hunk 3]
```diff
@@ -7344,6 +7551,14 @@ if (isUnsigned!F && isUnsigned!G && isUnsigned!H)
     return result;
 }
 
+///
+@safe pure nothrow @nogc unittest
+{
+    assert(powmod(1U, 10U, 3U) == 1);
+    assert(powmod(3U, 2U, 6U) == 3);
+    assert(powmod(5U, 5U, 15U) == 5);
```
Comment (wilzbach) [inline std/math.d]: Two or three more examples would be nice here, especially ones for which it gets clear that's `x^y % m`, e.g.

```d
    assert(powmod(2U, 3U, 5U) == 3);
    assert(powmod(2U, 4U, 5U) == 1);
    assert(powmod(2U, 5U, 5U) == 2);
```

File: std/math.d (hunk 2 above)
Comment (JackStouffer) [reply std/math.d]: forgot

File: std/math.d (hunk 1 above)
Comment (n8sh) [reply std/math.d]: `copysign` should have no imprecision. Using `feqrel` here would cause confusion by implying that it does.

File: std/math.d [hunk 4]
```diff
@@ -7204,6 +7391,29 @@ if (isFloatingPoint!(F) && isFloatingPoint!(G))
     return impl(x, y);
 }
 
+///
+@safe pure nothrow @nogc unittest
+{
+    assert(pow(1.0, 2.0).feqrel(1.0) > 16);
+    assert(pow(0.0, 0.0).feqrel(1.0) > 16);
```
Comment (n8sh) [inline std/math.d]: Exact equality, not `freqrel`.

File: std/math.d [hunk 5]
```diff
@@ -7204,6 +7391,29 @@ if (isFloatingPoint!(F) && isFloatingPoint!(G))
     return impl(x, y);
 }
 
+///
+@safe pure nothrow @nogc unittest
+{
+    assert(pow(1.0, 2.0).feqrel(1.0) > 16);
```
Comment (n8sh) [inline std/math.d]: Exact equality, not `freqrel`.

File: std/math.d [hunk 6]
```diff
@@ -5692,6 +5749,21 @@ private:
     }
 }
 
+///
+@safe unittest
+{
+    FloatingPointControl fpctrl;
+
+    fpctrl.rounding = FloatingPointControl.roundDown;
+    assert(lrint(1.5).feqrel(1.0) > 16);
```
Comment (n8sh) [inline std/math.d]: For demonstrating rounding we want exact equality, not `freqrel`. Same for the rest of this unittest.

File: std/math.d [hunk 7]
```diff
@@ -6774,16 +6903,46 @@ T nextafter(T)(const T x, const T y) @safe pure nothrow @nogc
  */
 real fdim(real x, real y) @safe pure nothrow @nogc { return (x > y) ? x - y : +0.0; }
 
-/****************************************
+///
+@safe pure nothrow @nogc unittest
+{
+    assert(fdim(2.0, 0.0).feqrel(2.0) > 16);
```
Comment (n8sh) [inline std/math.d]: Don't use `feqrel` here or in the other examples currently in this unittest.

File: std/math.d [hunk 8]
```diff
@@ -6774,16 +6903,46 @@ T nextafter(T)(const T x, const T y) @safe pure nothrow @nogc
...
+/**
  * Returns the larger of x and y.
  */
 real fmax(real x, real y) @safe pure nothrow @nogc { return x > y ? x : y; }
 
-/****************************************
+///
+@safe pure nothrow @nogc unittest
+{
+    assert(fmax(0.0, 2.0).feqrel(2.0) > 16);
```
Comment (n8sh) [inline std/math.d]: `fmax` is not imprecise. Do not use `feqrel` in these examples.

File: std/math.d [hunk 9]
```diff
@@ -6774,16 +6903,46 @@ T nextafter(T)(const T x, const T y) @safe pure nothrow @nogc
...
+
+/**
  * Returns the smaller of x and y.
  */
 real fmin(real x, real y) @safe pure nothrow @nogc { return x < y ? x : y; }
 
+///
+@safe pure nothrow @nogc unittest
+{
+    assert(fmin(0.0, 2.0).feqrel(0.0) > 16);
```
Comment (n8sh) [inline std/math.d]: `fmin` is not imprecise. Do not use `feqrel` in these examples.

File: std/math.d [hunk 10]
```diff
@@ -6792,7 +6951,17 @@ real fmin(real x, real y) @safe pure nothrow @nogc { return x < y ? x : y; }
  */
 real fma(real x, real y, real z) @safe pure nothrow @nogc { return (x * y) + z; }
 
-/*******************************************************************
+///
+@safe pure nothrow @nogc unittest
+{
+    assert(fma(0.0, 2.0, 2.0).feqrel(2.0) > 16);
```
Comment (n8sh) [inline std/math.d]: Because `fma` makes rounding guarantees use here exact equality, not `feqrel`.

File: std/math.d [hunk 11]
```diff
@@ -6843,6 +7012,15 @@ if (isFloatingPoint!(F) && isIntegral!(G))
     return p;
 }
 
+///
+@safe pure nothrow @nogc unittest
+{
+    assert(pow(2.0, 5).feqrel(32.0) > 16);
```
Comment (n8sh) [inline std/math.d]: Don't use `feqrel` for `pow(2.0, 5)`. Keep it for the other example.

File: std/math.d [hunk 12]
```diff
@@ -6961,7 +7139,16 @@ if (isIntegral!I && isFloatingPoint!F)
     return pow(cast(real) x, cast(Unqual!F) y);
 }
 
-/*********************************************
+///
+@safe pure nothrow @nogc unittest
+{
+    assert(pow(2, 5).feqrel(32.0) > 16);
+    assert(pow(7, 3).feqrel(343.0) > 16);
```
Comment (n8sh) [inline std/math.d]: Don't use `feqrel` for `pow(2, 5)` or `pow(7, 3)`.

File: std/math.d [hunk 13]
```diff
@@ -5045,6 +5045,35 @@ real remquo(real x, real y, out int n) @trusted nothrow @nogc  /// ditto
         assert(0, "remquo not implemented");
 }
 
+///
+@safe @nogc nothrow unittest
+{
+    assert(remainder(5.1, 3.0).feqrel(-0.9) > 16);
+    assert(remainder(-5.1, 3.0).feqrel(0.9) > 16);
+    assert(remainder(0.0, 3.0).feqrel(0.0) > 16);
```
Comment (n8sh) [inline std/math.d]: Use `feqrel` here only on the examples where the result is +/- 0.9.

File: std/math.d [hunk 14]
```diff
@@ -5045,6 +5045,35 @@ real remquo(real x, real y, out int n) @trusted nothrow @nogc  /// ditto
...
+    assert(remainder(1.0, real.infinity).feqrel(1.0) > 16);
+}
+
+///
+@safe @nogc nothrow unittest
+{
+    int n;
+
+    assert(remquo(5.1, 3.0, n).feqrel(-0.9) > 16 && n == 2);
+    assert(remquo(-5.1, 3.0, n).feqrel(0.9) > 16 && n == -2);
```
Comment (n8sh) [inline std/math.d]: Keep `feqrel` when the result is +/- 0.9. Remove it for other cases.

File: std/math.d [hunk 15]
```diff
@@ -5045,6 +5045,35 @@ real remquo(real x, real y, out int n) @trusted nothrow @nogc  /// ditto
...
 }
 
+///
+@safe @nogc nothrow unittest
+{
+    assert(remainder(5.1, 3.0).feqrel(-0.9) > 16);
+    assert(remainder(-5.1, 3.0).feqrel(0.9) > 16);
+    assert(remainder(0.0, 3.0) == 0.0);
+
+    assert(remainder(real.infinity, 3.0) is -real.nan);
```
Comment (wilzbach) [inline std/math.d]: this line is failing on the auto-tester:

```
core.exception.AssertError@std/math.d(5027): unittest failure
----------------
??:? _d_unittestp [0x2bb32f79]
??:? nothrow @nogc @safe void std.math.__unittest_L5018_C21() [0x2a067ca9]
??:? void std.math.__modtest() [0x2a0b18c2]
??:? ??? [0x804917a]
```

File: std/math.d [hunk 16]
```diff
@@ -5045,6 +5045,35 @@ real remquo(real x, real y, out int n) @trusted nothrow @nogc  /// ditto
...
+@safe @nogc nothrow unittest
+{
+    int n;
+
+    assert(remquo(5.1, 3.0, n).feqrel(-0.9) > 16 && n == 2);
+    assert(remquo(-5.1, 3.0, n).feqrel(0.9) > 16 && n == -2);
+    assert(remquo(0.0, 3.0, n) == 0.0 && n == 0);
+
+    // in special cases, n is set to zero
+    assert(remquo(real.infinity, 3.0, n) is -real.nan && n == 0);
```
Comment (wilzbach) [inline std/math.d]: this line is failing on the auto-tester:

```
core.exception.AssertError@std/math.d(5027): unittest failure
----------------
??:? _d_unittestp [0x2bb32f79]
??:? nothrow @nogc @safe void std.math.__unittest_L5018_C21() [0x2a067ca9]
??:? void std.math.__modtest() [0x2a0b18c2]
??:? ??? [0x804917a]
```

**Comment by JackStouffer** [comment]
https://issues.dlang.org/show_bug.cgi?id=18733

**Comment by wilzbach** [comment]
@JackStouffer we can't merge this as long as it's failing on FreeBSD :/ (so removing the 72h merge label for now)
Maybe `version` the failing test + put a FIXME bugzilla issue like we did for std.random?

**Comment by JackStouffer** [comment]
The most annoying thing about this is that the line numbers on the error messages in the auto tester are clearly wrong. So, I'm doing this manually to figure out what's going on.

**Comment by wilzbach** [comment]
just rebase your PR to master - then the line number will match ;-)

**Comment by JackStouffer** [comment]
This is what happens when you have no tests https://issues.dlang.org/show_bug.cgi?id=18739

File: std/math.d [hunk 17]
```diff
@@ -5054,6 +5054,33 @@ real remquo(real x, real y, out int n) @trusted nothrow @nogc  /// ditto
...
+@safe @nogc nothrow unittest
+{
+    version(Posix)
+    {
+        assert(remainder(5.1, 3.0).feqrel(-0.9) > 16);
+        assert(remainder(-5.1, 3.0).feqrel(0.9) > 16);
+        assert(remainder(0.0, 3.0) == 0.0);
+
+        assert(remainder(1.0, 0.0) is -real.nan);
+        assert(remainder(-1.0, 0.0) is -real.nan);
```
Comment (joakim-noah) [inline std/math.d]: These two tests assert for me on Android/ARM with the latest ldc 1.10 beta, as Bionic's `remainderl` returns `real.nan` instead. @kinke, same on linux/armhf?

@JackStouffer, where did you get the values from the table above, wouldn't seem to matter what the sign is, as the result is undefined?

**Pull request merged**